### ISSUE TITLE
+ letter in s3 bucket path

### ISSUE THREAD

**Issue by bintriz** [samtools/samtools#775]
The aws client can handle + letter in s3 path. But, samtools can't. I had to change the + letter into the url encoding %2b.

Example
```
[ec2-user@ip-172-31-1-20 20.1st_validation]$ aws s3 ls s3://nda-bsmn-scratch/3340241/f8d7056e-a875-4186-87ad-87747b2ac964/5154_NeuN_positive-BSMN_REF_NeuN+-534-U01MH106876.bam
2018-01-03 18:59:27 200319335583 5154_NeuN_positive-BSMN_REF_NeuN+-534-U01MH106876.bam
[ec2-user@ip-172-31-1-20 20.1st_validation]$ samtools view -H s3://nda-bsmn-scratch/3340241/f8d7056e-a875-4186-87ad-87747b2ac964/5154_NeuN_positive-BSMN_REF_NeuN+-534-U01MH106876.bam|head
[E::hts_open_format] Failed to open file s3://nda-bsmn-scratch/3340241/f8d7056e-a875-4186-87ad-87747b2ac964/5154_NeuN_positive-BSMN_REF_NeuN+-534-U01MH106876.bam
samtools view: failed to open "s3://nda-bsmn-scratch/3340241/f8d7056e-a875-4186-87ad-87747b2ac964/5154_NeuN_positive-BSMN_REF_NeuN+-534-U01MH106876.bam" for reading: No such file or directory
[ec2-user@ip-172-31-1-20 20.1st_validation]$ samtools view -H s3://nda-bsmn-scratch/3340241/f8d7056e-a875-4186-87ad-87747b2ac964/5154_NeuN_positive-BSMN_REF_NeuN%2b-534-U01MH106876.bam|head
@HD	VN:1.5	GO:none	SO:coordinate
@SQ	SN:1	LN:249250621
@SQ	SN:2	LN:243199373
@SQ	SN:3	LN:198022430
@SQ	SN:4	LN:191154276
@SQ	SN:5	LN:180915260
@SQ	SN:6	LN:171115067
@SQ	SN:7	LN:159138663
@SQ	SN:8	LN:146364022
@SQ	SN:9	LN:141213431
```

**Comment by daviesrob** [comment]
Presumably the aws client is percent-encoding the `+` sign for you when it converts the s3 path into the https:// URL used to access the data.  We could implement that in the s3:// handler used by samtools as well, although it would have the side-effect of breaking any s3 paths that have already been percent-encoded.

**Comment by dkj** [comment]
`#` is also a problem - it will work if I replace them with `%23` but that is inconsistent with other s3 URL capable tools I think.

**Comment by jmarshall** [comment]
If the aws client is doing some percent-encoding of user-supplied URLs, then that's a bit of a hack.

Typically a client and certainly a client API function wouldn't attempt to percent-encode a user-supplied URL itself (compare [libcurl's function](https://curl.haxx.se/libcurl/c/CURLOPT_URL.html)), as in general it's impossible to tell whether a particular character is desired to be a metacharacter or not — and that way leads to madness.

So this code was written intentionally not to try to percent-encode the URL on the user's behalf.

For example (assuming the [versioning described in the S3 documentation](https://docs.aws.amazon.com/AmazonS3/latest/API/RESTObjectGET.html#RESTObjectGET-versioning) is still a thing), blanket percent-encoding of user-supplied URLs will make it difficult to specify query parameters when the user needs them:

    s3://bucket/myObject?versionId=3/L4kqtJlcpXroDTDmpUMLUo

**Comment by whitwham** [comment]
The users need it percent encoding or it will not work.

**Comment by daviesrob** [comment]
The aws client does percent-encode.  It's a bit difficult to find, but I think it's [here](https://github.com/boto/botocore/blob/eb474884/botocore/serialize.py#L446).  s3cmd also does something similar, and has a `verbatim` option to turn off the encoding.

The problem is that people expect S3 to work more like a filesystem than a web server, and get confused when names with things like `+` or `#` in fail to work as expected.  In one spectacular case here it caused no less than 500,000 files to be reported as missing when in fact they were (mostly) perfectly fine...

**Comment by whitwham** [comment]
Should be fixed in PR samtools/htslib#839.
